task/copy.go

```diff
@@ -1,336 +1,135 @@
-	bytesPool := &sync.Pool{
-		New: func() interface{} {
-			return bytes.NewBuffer(make([]byte, 0, t.GetPartSize()))
-		},
-	}
-	t.SetBytesPool(bytesPool)
-}
-func (t *CopyStreamTask) run(ctx context.Context) error {
-	initTask := NewSegmentInit(t)
-	err := utils.ChooseDestinationStorageAsIndexSegmenter(initTask, t)
-	if err != nil {
-		return err
-	}
+func (a *Agent) HandleCopySingleFile(ctx context.Context, msg protobuf.Message) error {
+	log := zapcontext.From(ctx)
 
-	// TODO: we will use expect size to calculate part size later.
-	// if part size not set, use default value, otherwise load part size automatically
-	if !t.ValidatePartSize() {
-		t.SetPartSize(constants.DefaultPartSize)
-	}
+	arg := msg.(*proto.CopySingleFile)
 
-	if err := t.Sync(ctx, initTask); err != nil {
-		return err
-	}
-	t.SetSegment(initTask.GetSegment())
+	log.Info("copy single file",
+		zap.String("from", arg.SrcPath),
+		zap.String("to", arg.DstPath))
+	return nil
+}
+func (a *Agent) HandleCopyMultipartFile(ctx context.Context, msg protobuf.Message) error {
+	log := zapcontext.From(ctx)
 
-	offset, part := int64(0), 0
-	for {
-		it := NewSegmentStreamInit(t)
-		if err := t.Sync(ctx, it); err != nil {
-			return err
-		}
+	arg := msg.(*proto.CopyMultipartFile)
 
-		x := NewCopyPartialStream(t)
-		x.SetSize(it.GetSize())
-		x.SetContent(it.GetContent())
-		x.SetOffset(offset)
-		x.SetIndex(part)
-		t.Async(ctx, x)
+	// Send task and wait for response.
+	log.Info("copy multipart",
+		zap.String("from", arg.SrcPath),
+		zap.String("to", arg.DstPath))
+	return nil
+}
 
-		if it.GetDone() {
-			break
-		}
+func (a *Agent) HandleCopyMultipart(ctx context.Context, msg protobuf.Message) error {
+	log := zapcontext.From(ctx)
 
-		offset += it.GetSize()
-		part++
-	}
+	arg := msg.(*proto.CopyMultipart)
 
-	// if meet error, not continue run complete task
-	if err := t.Await(); err != nil {
-		return err
+	src := a.storages[arg.Src]
+	dst := a.storages[arg.Dst]
+	mulipart, ok := dst.(types.Multiparter)
+	if !ok {
+		log.Warn("storage does not implement Multiparter",
+			zap.String("storage", dst.String()))
+		return fmt.Errorf("not supported")
 	}
-	return t.Sync(ctx, NewSegmentCompleteTask(initTask))
-}
 
-func (t *CopyPartialStreamTask) new() {}
-func (t *CopyPartialStreamTask) run(ctx context.Context) error {
-	tk := token.FromContext(ctx)
-	tk.Take()
-	defer tk.Return()
+	r, w := iowrap.Pipe()
 
-	copyTask := NewSegmentStreamCopy(t)
-	if t.GetCheckMD5() {
-		md5sumTask := NewMD5SumStream(t)
-		if err := t.Sync(ctx, md5sumTask); err != nil {
-			return err
+	go func() {
+		o := dst.Create(arg.DstPath, ps.WithMultipartID(arg.MultipartId))
+		_, err := mulipart.WriteMultipart(o, r, arg.Size, int(arg.Index))
+		if err != nil {
+			log.Error("write multipart", zap.Error(err))
 		}
-		copyTask.SetMD5Sum(md5sumTask.GetMD5Sum())
-	}
+	}()
 
-	err := utils.ChooseDestinationStorageAsDestinationIndexSegmenter(copyTask, t)
+	_, err := src.Read(arg.SrcPath, w)
```

Comment:
Add pairs `offset` and `size` in `Read`?

Reply:
We will add them in following PRs.